transcom/mymove
# INT B-19939 Diversion Reason in API

## [B-19939](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A948732)

## Summary

This ticket adds a field for a diversion request's reason to the API. The field can be edited by the request diversion endpoint and viewed in any endpoint returning a full MTO Shipment. Generally followed how rejection reason was added to the reject shipment endpoint.

Upstream Dependency: [B-19938](https://github.com/transcom/mymove/pull/12589)

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
I'd recommend testing in both the app, to ensure the existing request diversion button still works, and postman or swagger, to ensure the endpoint works fully as intended.

#### In the App
1. Create or Identify an approved HHG to request diversions on (PENDNG & BILWEI should work if they are fresh).
2. Access the office app.
3. Login as a TOO.
4. Select one of the moves identified above from your queue.
5. Navigate to the move task order page and locate the request diversion button.
6. Open the developer tools' network tab to capture the request we're about to send.
7. Click the request diversion button, the UI should show the request was made successfully.
8. Find the request named "request-diversion"
9. The request's payload should have a body containing a blank "diversionReason"
10. The response should be a 200 and include`status: "DIVERSION_REQUESTED"` and `diversionReason: ""`. The diversion reason is blank so that the front-end can still send the request since a body is now required. The field will be made to require a non-blank string in a later ticket when the user is able to submit a reason.

#### In Postman
1. Create or Identify an approved HHG to request diversions on (PENDNG & BILWEI should work if they are fresh). 
2. For authentication to use in Postman, access the office app and login as a TOO.
3. Obtain the ID and eTag of the shipment you intend to request as well as your the x-Csrf-Token
4. Submit a POST request to the following URL with the following body
URL:`http://officelocal:3000/ghc/v1/shipments/{shipmentID}/request-diversion`
Body `{"diversionReason": "Your diversion reason"}`
5. The response should include `status: "DIVERSION_REQUESTED"` and `diversionReason: "Your diversion reason"`
6. Optionally verify the reason you submitted is in the database in the mto_shipments.diversion_reason column.

### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
